src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -1369,13 +1409,49 @@ class ArrayGenerator : public Generator<bsoncxx::array::value> {
           _node{node},
           _generatorArgs{generatorArgs},
           _valueGen{valueGenerator<false, UniqueAppendable>(node["of"], generatorArgs, parsers)},
-          _nTimesGen{intGenerator(extract(node, "number", "^Array"), generatorArgs)} {}
+          _nTimesGen{intGenerator(extract(node, "number", "^Array"), generatorArgs)},
+          _distinct {node["distinct"] ? node["distinct"].maybe<bool>().value_or(false) : false} {}
 
     bsoncxx::array::value evaluate() override {
         bsoncxx::builder::basic::array builder{};
         auto times = _nTimesGen->evaluate();
-        for (int i = 0; i < times; ++i) {
-            _valueGen->append(builder);
+        if (_distinct) {
+            // Choose initial bucket size of 2N to minimize hash collisions and resizes.
+            std::unordered_set<bsoncxx::array::value, SetHasher, SetKeyEq> distinctValues(2 * times);
+            int8_t consecutiveRepeats = 0;
+            const int8_t consecutiveRepeatsThreshold = 100;
+
+            while (distinctValues.size() < times) {
+                bsoncxx::builder::basic::array sbuilder{};
+                _valueGen->append(sbuilder);
+                auto [itr, succeeded] = distinctValues.insert(sbuilder.extract());
+
+                if (succeeded) {
+                    auto& element = *itr->view().begin();
+                    // Stolen shamelessly from the CXX driver library (bsoncxx/types/bson_value/view.cpp).
+                    switch (static_cast<int>(element.type())) {
```

Comment:
Is the \#define and \#include come from the shamelessly stolen part? Maybe just given an end line number or add a comment "end stolen shamelessly"